ITS alignment monitoring (check with QCv20240903), revised 20240919

delico18 opened the pull request:
New 14 plots(7, layer x 2, axis[z \~ alpide col, phi \~ alpide row] for ITS alignment monitoring are added.

- Circle Fitter (in XY plane)
- Line Fitter (in DZ plane), D is almost same as the path length of particle.

Based on Least Square Fit, we can observe the residual of particles' track, which we believe they will show the alignment quality of ITS.

iravasen commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -92,6 +111,10 @@ void ITSTrackTask::monitorData(o2::framework::ProcessingContext& ctx)
 
+  //o2::base::GeometryManager::loadGeometry();
+  mGeom = o2::its::GeometryTGeo::Instance();
+  mGeom->fillMatrixCache(o2::math_utils::bit2Mask(o2::math_utils::TransformType::L2G));
```
move these two lines inside the `if(mTimestamp == -1)`.

iravasen replied:
In addition here you are using the full geometry. I guess a light-weight version of it should be fine. Please copy what was done for the ClusterTask here: https://github.com/AliceO2Group/QualityControl/pull/2164/files#diff-41e35aa0ad16f0bc0150242f13cb67d78b78f53b020364e33f18cce3fcf44de4R125 

Note also the removal from the json of the `grpGeomRequest`

iravasen commented on Modules/ITS/itsTrack.json:
```diff
@@ -45,9 +45,24 @@
+          "MomResParMSC2": "25, 25, 25, 75, 75, 75, 75",
+          "ResidualMonitorTrackMinPt" : "0.00"
+        },
+	"grpGeomRequest" : {
```
`grpGeomRequest` can be removed if you use the light weight version of the geometry. See the other comment.

iravasen commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -628,6 +723,21 @@ void ITSTrackTask::createAllHistos()
+  for (int l = 0; l < NLayer; l++) {
+    //sensor
+    hdxySensor[l] = std::make_unique<TH2D>(Form("hdxy%dSensor",l), Form("ChipID vs dxy, Layer %d",l), 
+                                           1000, -0.05, 0.05, ChipBoundary[l+1] - ChipBoundary[l], ChipBoundary[l], ChipBoundary[l+1]);
```
do you really need 1000 bins on the X axis? Same comment for `hdz` histo. Bin width is 1 um like this, maybe 5 um are enough (i.e. comparable with the ALPIDE spatial reso).

iravasen replied:
Make the y axis bin with the chipID in the center: 
`ChipBoudary[l+1] - ChipBoundary[l] + 1, ChipBoundary[l] - 0.5, ChipBoundary[l+1] - 0.5`

Note the -0.5 also for the upper limit because for example chip 108 belongs already to L1 (and not to L0) if we start counting chipID from 0. Maybe make again the calculation but it should be correct.

iravasen commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -648,4 +758,455 @@ void ITSTrackTask::publishHistos()
+    }
+  }
+  
+  /*
```
remove this long comment.

iravasen commented on Modules/ITS/include/ITS/ITSTrackTask.h:
```diff
@@ -118,6 +125,147 @@ class ITSTrackTask : public TaskInterface
+  std::array<std::unique_ptr<TH2D>, NLayer> hdxySensor{};//[NLayer];
+  std::array<std::unique_ptr<TH2D>, NLayer> hdzSensor{};//[NLayer];
+
+  void circlefit_XY(double* input, double* par, double &MSEvalue, std::vector<bool> hitUpdate, int step = 0);
```
use the camel case for all functions: `circlefit_XY` becomes `circleFitXY`. This is the convention in QC.

iravasen commented:
I will apply the clang format at the end once all comments have been implemented

iravasen commented on Modules/ITS/itsTrack.json:
```diff
@@ -45,9 +45,24 @@
           "doNorm" : "1",
-          "InvMasses" : "0"
+          "InvMasses" : "0",
+          "UseDefaultMomResPar" : "0",
```
Can you please add an option to enable/disable your new plots? Default is 1 (i.e. enabled).

delico18 commented:
2024 09 19 update

There seems to be tiny errors in getting layer of cluster, and I corrected the lines.

Options...
to activate alignment monitoring(itsTrack.json) is added.
to request the minimum number of clusters / track for alignment monitoring(itsTrack.json) is added.

I updated the simple version of documentation that describes the fitting & residual calculation procedures.

iravasen commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -628,6 +727,21 @@ void ITSTrackTask::createAllHistos()
   hTrackPtVsPhi->SetStats(0);
+
+  for (int l = 0; l < NLayer; l++) {
+    //sensor
```
Initialize histograms only `if (mAlignmentMonitor == 1)`

iravasen commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -441,6 +535,11 @@ void ITSTrackTask::reset()
   hTrackPtVsPhi->Reset();
+
+  for (int l = 0; l < NLayer; l++) {
+    hdxySensor[l]->Reset();
```
Reset histo only `if mAlignmentMonitor == 1`

IsakovAD commented on Modules/ITS/include/ITS/ITSTrackTask.h:
```diff
@@ -118,6 +125,149 @@ class ITSTrackTask : public TaskInterface
+  //analysis for its-only residual
+  o2::its::GeometryTGeo* mGeom;
+
+  double FitStepSize0 = 0.3;
+  double FitStepSize1 = 1.0e-5;
+  double FitStepSize2 = 1.0e-5;
+  double FitStepSize3 = 1.0e-5;
```
Could allocate a vector for that?

IsakovAD commented on Modules/ITS/include/ITS/ITSTrackTask.h:
```diff
@@ -118,6 +125,149 @@ class ITSTrackTask : public TaskInterface
+
+  TH1D* hResidualRealTimePerTrackwFinerBin;
+  TH1D* hResidualCPUTimePerTrackwFinerBin;
+  std::array<std::unique_ptr<TH2D>, NLayer> hdxySensor{};//[NLayer];
+  std::array<std::unique_ptr<TH2D>, NLayer> hdzSensor{};//[NLayer];
```
Do you that that we need TH2D in this case? The general issue which we are dealing with is the size of objects. Could you try configuration with TH2S or TH2I. In my understanding, you have very fine binning and we do not expect to have so many entries in each bin

delico18 replied:
I changed this to TH2I, but I think TH2F can be more generally used in case we just take care of the size of objects.
Because the size of TH2I and TH2F is the same, but the maximum bin content for TH2F is more large.

If there is no problem using TH2F, I'd like to choose TH2F. Please leave your comments.

IsakovAD commented on Modules/ITS/include/ITS/ITSTrackTask.h:
```diff
@@ -118,6 +125,149 @@ class ITSTrackTask : public TaskInterface
+
+  void lineFitDZ(double* zIn, double* betaIn, double* parz, double Radius, bool vertex, std::vector<bool> hitUpdate);
+
+  double sigma_meas[2][NLayer] = {{45,45,45,55,55,55,55},
```
Is it a duplicate on of line 165?

delico18 replied:
No, one is data member of ITSTrackTask, and another one is data member of struct "se_circlefitXY".

I renamed the former case as "mSigmaMeas, mSigmaMsc" to have similar patterns for the normal data member in ITSTrackTask.

IsakovAD commented on Modules/ITS/src/ITSTrackTask.cxx:
```diff
@@ -100,6 +121,11 @@ void ITSTrackTask::monitorData(o2::framework::ProcessingContext& ctx)
     mDict = TaskInterface::retrieveConditionAny<o2::itsmft::TopologyDictionary>("ITS/Calib/ClusterDictionary", metadata, ts);
     ILOG(Debug, Devel) << "Dictionary size: " << mDict->getSize() << ENDM;
+
+    o2::its::GeometryTGeo::adopt(TaskInterface::retrieveConditionAny<o2::its::GeometryTGeo>("ITS/Config/Geometry", metadata, ts));
```
If we do not use geometry in the code besides alignment monitoring, could you make geometry downloading optional?

delico18 replied:
It is implemented in the code and failure case(throw) to download ccdb object for ITS geometry is added.

IsakovAD commented:
Hello, 

Nice new feature! General comments: 
1) Size of new objects: I assume that TH2D is too much. Could you please check the size of objects in case of TH2D and TH2S/TH2I. I think that with such nice binning double is a quite overkill 
2) Try to minimize influence of your code when AlignmentMonitoring set to FALSE: do not initialize objects/do not download geometry etc. Please, check that task still working if user will set monitoring to OFF.

 Otherwise,  looking forward to results of this new feature

delico18 commented:
The monitoring objects for residuals are only activated when mAlignmentMonitor is true.
There is no more influence related to this object when this is set to false.

Please review this PR.(including TH2F vs TH2I)
